servo/events.py

```diff
@@ -518,11 +518,14 @@ def decorator(fn: EventCallable) -> EventCallable:
         handler_mod_name = handler_localns.get("__module__", None)
         handler_module = sys.modules[handler_mod_name] if handler_mod_name else None
 
-        if preposition == Preposition.before:
-            before_handler_signature = inspect.Signature.from_callable(__before_handler)
+        if preposition in (Preposition.before, Preposition.on):
+            ref_signature=event.signature
+            if preposition == Preposition.before:
+                # 'before' event takes same args as 'on' event, but returns None
+                ref_signature = ref_signature.replace(return_annotation="None")
             servo.utilities.inspect.assert_equal_callable_descriptors(
                 servo.utilities.inspect.CallableDescriptor(
-                    signature=before_handler_signature,
+                    signature=ref_signature,
```

Comment:
What has motivated this change to the before handlers? It has undermined the entire point of the handler chain -- to allow the set of enabled connectors to collaborate through loose coupling.

Adding `Control` as a required input to all event handlers is one thing but now by collapsing the before and on handlers to having the same signature, it is impossible to program before handlers that function abstractly against any event.

This undermines a key part of the design. If you think about use-cases like enforcing a security or timing policy against all operations (e.g., the blackout policy defined by Ancestry), what you want to be able to do is define a single filter callback that applies against all the events. Knowing about each event and defining handlers for each signature breaks this API contract.

What I can get behind is using `*args` and `**kwargs` smartly to allow you to get access to these values if you need them and are programming against a specific event.

Reply:
> it is impossible to program before handlers that function abstractly against any event.

Not really, just add `*args, **kwargs` to the signature of the generic handler and the validation will allow it through. If we're programming against a specific event then we should be able to explicitly type out the parameters expected for that event and have validation in place to ensure those match the event being handled. Using `*args, **kwargs` for specific events seems less robust as the event signature could change which would raise a less informative KeyError during runtime instead of being caught on initialization...